### ISSUE TITLE
False positive for reportAttributeAccessIssue

### ISSUE THREAD

**Issue by trinhanhngoc** [microsoft/pyright#10044]
Pyright 1.1.396

Code sample to reproduce the issue:

```python
def xxx(x: int) -> int:
    return x

class XXX:
    x = xxx

print(XXX.x(1)) # reportAttributeAccessIssue
```

Error:

```
Cannot access attribute "x" for class "type[XXX]"
  Could not bind method "xxx" because "XXX" is not assignable to parameter "x"
    "XXX" is not assignable to "int"Pyright[reportAttributeAccessIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportAttributeAccessIssue)
(variable) x: Unknown
```

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

In your code sample, `x` is an instance method because it is not decorated with `@staticmethod`. That means the first parameter is considered a `self` parameter. In the expression `XXX.x`, the method is bound to the class `XXX` and partially specialized. During this operation, the type checker validates that `XXX` is assignable to the declared type of the first (i.e. the `self`) parameter, and in this case it is not. Binding and partial specialization is especially important when the class is generic and the `self` parameter is annotated with a type that includes a type variable, but it's also important for non-generic classes like in your example.

You can fix the bug in your code a couple of ways:

1. Mark `xxx` as a staticmethod.
```python
class XXX:
    x = staticmethod(xxx)
```

or

```python
@staticmethod
def xxx(x: int) -> int:
    return x
```

2. Create a separate static method in XXX that calls through to the other `xxx` function:

```python
class XXX:
    @staticmethod
    def x(x: int) -> int:
        return xxx(x)
```

While option 2 is a bit more verbose, I think it's the technique I'd use because it's the easiest to read and maintain.

**Comment by trinhanhngoc** [comment]
@erictraut ,

Thank you for the detailed explanation. However, my code sample works perfectly at runtime and has no errors with mypy. It will be great if Pyright also behaves the same.

**Comment by erictraut** [comment]
Mypy does not properly perform method binding in conformance with the Python typing spec, so this a bug in mypy not in pyright. Prior to [PEP 696](https://peps.python.org/pep-0696/), the binding rules were unclear in the typing spec, but this PEP makes it clear that specialization and binding must occur when accessing a method even if it's an instance method accessed from a class object.
